### PR TITLE
native: add native periph timer

### PR DESCRIPTION
**Pull request by kaspar030**
This PR adds a simple wrapper from periph/timer.h to host itimers.

It is based un @LudwigOrtmann hwtimer implementation. I've dropped the multiplexing, though.

One of the commits marks some functions in ```hwtimer.c``` ```static```, as one of them was colliding with ```timer.c```.

### CODE REVIEW

**Comment by LudwigOrtmann** [comment]
PR needs rebase for Travis mirror setup

**Comment by kaspar030** [comment]
rebased.

**Comment by kaspar030** [comment]
I'm aware that this will collide with the native hwtimer implementation. But the (soon to come) new timer subsystem will drop hwtimer, and I'd like native users to be able to test...

**Comment by LudwigOrtmann** [comment]
Maybe catch this case in make?

**Comment by kaspar030** [comment]
Not so easy as hwtimer is still a core dependency with lots of uses.
But in my tests, just using this native periph timers works fine for testing timers. ;)

**Comment by LudwigOrtmann** [comment]
Print out a big red blinking warning instead of failing then..

**Comment by kaspar030** [comment]
Could we also use the hwtimer wrapper used by some platforms to keep compatibility?

**Comment by LudwigOrtmann** [comment]
I don't know about the details, but I'd actually prefer that if it is what I think it is.

**Comment by kaspar030** [comment]
Some platforms wrap the hwtimer interface around one periph timer so as to fulfill the core hwtimer dependancies. We could do the same here. Will do shortly.

**Comment by kaspar030** [comment]
- rebased (was missing vtimer test fixes for review)
- native now uses the slim periph timer + hwtimer_compat (compare #2913)

**Comment by LudwigOrtmann** [comment]
This PR includes parts of #2913 now?

**Comment by kaspar030** [comment]
uff, yes. I didn't want to just copy the hwtimer->periph wrapper. So I made it a module and PR'ed it. It's very simple, and Hauke already ACKed it, but yes, I cherry-picked it in here.

**Comment by LudwigOrtmann** [comment]
I just wondered if this PR needs a label indicating its dependency on another PR.

**Comment by kaspar030** [comment]
- rebased,  #2913 is in master now

**Comment by LudwigOrtmann** [comment]
```
/home/lo/RIOT/sys/compat/hwtimer/hwtimer_arch.c: In function ‘hwtimer_arch_init’:
/home/lo/RIOT/sys/compat/hwtimer/hwtimer_arch.c:36:55: warning: unused parameter ‘fcpu’ [-Wunused-parameter]
 void hwtimer_arch_init(void (*handler)(int), uint32_t fcpu)
                                                       ^
```

File: cpu/native/periph/timer.c [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    mach_timespec_t mts;
+    host_get_clock_service(mach_host_self(), SYSTEM_CLOCK, &cclock);
+    clock_get_time(cclock, &mts);
+    mach_port_deallocate(mach_task_self(), cclock);
+    t.tv_sec = mts.tv_sec;
+    t.tv_nsec = mts.tv_nsec;
+#else
+
+    if (real_clock_gettime(CLOCK_MONOTONIC, &t) == -1) {
+        err(EXIT_FAILURE, "hwtimer_arch_now: clock_gettime");
```
Comment (LudwigOrtmann) [inline cpu/native/periph/timer.c]: `s/hwtimer_arch_now/timer_read/`

File: cpu/native/periph/timer.c [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+int timer_clear(tim_t dev, int channel)
+{
+    (void)dev;
+    (void)channel;
+    struct itimerval null_timer = {{0}, {0}};
+    _native_syscall_enter();
+    if (real_setitimer(ITIMER_REAL, &null_timer, NULL) == -1) {
+        err(EXIT_FAILURE, "schedule_timer: setitimer");
```
Comment (LudwigOrtmann) [inline cpu/native/periph/timer.c]: `s/schedule_timer/timer_clear/`

File: cpu/native/periph/timer.c [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+    struct itimerval itv = {{0}, {0}};
+    itv.it_value.tv_sec = (offset / 1000000);
+    itv.it_value.tv_usec = offset % 1000000;
+
+    DEBUG("timer_set(): setting %u.%06u\n", (unsigned)itv.it_value.tv_sec, (unsigned)itv.it_value.tv_usec);
+
+    _native_syscall_enter();
+    if (real_setitimer(ITIMER_REAL, &itv, NULL) == -1) {
+        err(EXIT_FAILURE, "schedule_timer: setitimer");
```
Comment (LudwigOrtmann) [inline cpu/native/periph/timer.c]: `s/schedule_timer/timer_set/`

**Comment by kaspar030** [comment]
- rebased, squashed

**Comment by LudwigOrtmann** [comment]
@kaspar030 needs rebase

**Comment by LudwigOrtmann** [comment]
Seems to fix https://github.com/RIOT-OS/RIOT/issues/1281

**Comment by LudwigOrtmann** [comment]
Possibly also https://github.com/RIOT-OS/RIOT/issues/715

**Comment by kaspar030** [comment]
@LudwigOrtmann should we merge?

**Comment by LudwigOrtmann** [comment]
I didn't have time to test on OS X or FreeBSD yet..

**Comment by kaspar030** [comment]
Could anyone test at least on OS X?

**Comment by LudwigOrtmann** [comment]
@josephnoir could you?

**Comment by emmanuelsearch** [comment]
I could also try to test this on OS X if someone explains how I should test ;)

**Comment by kaspar030** [comment]
@emmanuelsearch If the vtimer_msg test works, everything should be fine!

**Comment by emmanuelsearch** [comment]
mmmh define working. It seems to hang after some time (no consistent duration after which it hangs though). Below an example of what I get (but it can also happen that it hangs sooner, e.g. right after the first message is received in the second thread):
```
kernel_init(): This is RIOT! (Version: 2014.12-1718-gd3dadd-native-timer)
kernel_init(): jumping into first task...
This is thread 3
sending 1st msg
now=0:79 -> every 2.0s: Hello World
sending 2nd msg
now=0:147 -> every 5.0s: This is a Test
This is thread 4
sec=1 min=0 hour=0
now=2:308 -> every 2.0s: Hello World
sec=2 min=0 hour=0
sec=3 min=0 hour=0
now=4:2115 -> every 2.0s: Hello World
sec=4 min=0 hour=0
now=5:3641 -> every 5.0s: This is a Test
sec=5 min=0 hour=0
now=6:2658 -> every 2.0s: Hello World
sec=6 min=0 hour=0
sec=7 min=0 hour=0
now=8:4536 -> every 2.0s: Hello World
sec=8 min=0 hour=0
sec=9 min=0 hour=0
now=10:8861 -> every 5.0s: This is a Test
```
(and then it hangs)

**Comment by kaspar030** [comment]
You might be hitting the vtimer bugs. Can you check the master branch?

**Comment by emmanuelsearch** [comment]
Master does not compile right now on OS X. Below the error I get:
```
RIOT/core/include/thread.h:138:14: error: conflicting
      types for 'thread_create'
kernel_pid_t thread_create(char *stack,
             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/usr/include/mach/task.h:193:15: note: 
      previous declaration is here
kern_return_t thread_create
              ^
1 error generated.
```

**Comment by LudwigOrtmann** [comment]
@emmanuelsearch Please open an issue for this problem including your system and toolchain version.

**Comment by emmanuelsearch** [comment]
done https://github.com/RIOT-OS/RIOT/issues/3175
@LudwigOrtmann can you label it as you see fit?

**Comment by kaspar030** [comment]
- rebased
- fixed OS X compile, see #3288

**Comment by kaspar030** [comment]
@emmanuelsearch please retest!

**Comment by emmanuelsearch** [comment]
Does not compile on OS X as is. Will PR a small change that makes it compile later today.

Still hangs after random amount of time (sometimes seems to run normally).
If merged with #3288 it consistently hangs as show below.

```
kernel_init(): jumping into first task...
This is thread 3
sending 1st msg
now=0:97 -> every 2.0s: Hello World
sending 2nd msg
now=0:156 -> every 5.0s: This is a Test
This is thread 4
sec=1 min=0 hour=0
now=2:5457 -> every 2.0s: Hello World
```

should I try merging/testing something else?

**Comment by kaspar030** [comment]
@emmanuelsearch Does master behave differently for that test?

**Comment by OlegHahm** [comment]
I meant this one: master is not even building on OS X,b ut with #3288 and #3290, the tests behaved more or less the same IIRC. @emmanuelsearch, might correct me.

**Comment by kaspar030** [comment]
IMHO, if this is as good or better as master, we should merge and wait for improvements by #3282.

**Comment by OlegHahm** [comment]
:+1: Let's wait what @emmanuelsearch says.

**Comment by emmanuelsearch** [comment]
Still does not work as well as on master...
Hangs often (like 75% of the time) after a few seconds (between 2 and 6)
Current master never failed in my last round of tests.

**Comment by kaspar030** [comment]
Actually this is obsoleted by #3282, master contains this.

**Comment by emmanuelsearch** [comment]
great ;)

**Pull request closed without merging**